src/sgx-verifier/AttestationVerifier.sol

```diff
@@ -0,0 +1,207 @@
+        attestationVerifier = IDcapAttestation(_attestationVerifierAddr);
+    }
+
+    /*************************
+     * Public View Functions *
+     *************************/
+
+    /// @inheritdoc IAttestationVerifier
+    function isTrustedMrSigner(bytes32 mrSigner) public view returns (bool) {
+        return trustedUserMrSigner.contains(mrSigner);
+    }
+
+    /// @inheritdoc IAttestationVerifier
+    function isTrustedMrEnclave(bytes32 mrEnclave) public view returns (bool) {
+        return trustedUserMrEnclave.contains(mrEnclave);
+    }
+
+    /// @inheritdoc IAttestationVerifier
+    function getTrustedMrSigners() external view returns (bytes32[] memory signers) {
+        uint256 length = trustedUserMrSigner.length();
+        signers = new bytes32[](length);
+        for (uint256 i = 0; i < length; i++) {
+            signers[i] = trustedUserMrSigner.at(i);
+        }
+    }
+
+    /// @inheritdoc IAttestationVerifier
+    function getTrustedMrEnclaves() external view returns (bytes32[] memory enclaves) {
+        uint256 length = trustedUserMrEnclave.length();
+        enclaves = new bytes32[](length);
+        for (uint256 i = 0; i < length; i++) {
+            enclaves[i] = trustedUserMrEnclave.at(i);
+        }
+    }
+
+    /// @inheritdoc IAttestationVerifier
+    function verifyAttestation(bytes calldata _report, bytes32 _userData) external view {
+        (bool success, bytes memory output) = attestationVerifier.verifyAndAttestOnChain(_report);
+        if (!success) {
+            revert ErrorInvalidReport();
+        }
+
+        (bytes32 reportUserData, bytes32 mrEnclave, bytes32 mrSigner) = extractEnclaveReport(output);
+        if (reportUserData != _userData) {
+            revert ErrorReportDataMismatch();
+        }
+        // check local enclave report
+        if (!isTrustedMrEnclave(mrEnclave)) {
+            revert ErrorInvalidMrEnclave();
+        }
+        if (!isTrustedMrSigner(mrSigner)) {
+            revert ErrorInvalidMrSigner();
+        }
+    }
+
+    /************************
+     * Restricted Functions *
+     ************************/
+
+    /// @notice Update the status of a signer.
+    /// @param _mrSigner The signer to update.
+    /// @param _status The new status to update. If it is `true`, the signer is trusted.
+    function updateMrSigner(bytes32 _mrSigner, bool _status) external onlyOwner {
+        // @note No need to check the previous status, offline owner will make sure this tx is meaningful.
+        if (_status) {
+            trustedUserMrSigner.add(_mrSigner);
+        } else {
+            trustedUserMrSigner.remove(_mrSigner);
+        }
+
+        emit UpdateMrSigner(_mrSigner, _status);
+    }
+
+    /// @notice Update the status of an enclave.
+    /// @param _mrEnclave The enclave to update.
+    /// @param _status The new status to update. If it is `true`, the enclave is trusted.
+    function updateMrEnclave(bytes32 _mrEnclave, bool _status) external onlyOwner {
```

Comment:
Same question for MrEnclave.

Reply:
answered in https://github.com/scroll-tech/scroll-contracts/pull/58#discussion_r1828794676